distributed/joblib.py

```diff
@@ -45,31 +46,85 @@ def joblib_funcname(x):
     return funcname(x)
 
 
+class Batch(object):
+    def __init__(self, tasks):
+        self.tasks = tasks
+
+    def __call__(self, *data):
+        results = []
+        for func, args, kwargs in self.tasks:
+            args = [a(data) if isinstance(a, itemgetter) else a
+                    for a in args]
+            kwargs = {k: v(data) if isinstance(v, itemgetter) else v
+                      for (k, v) in kwargs.items()}
+            results.append(func(*args, **kwargs))
+        return results
+
+    def __reduce__(self):
+        return (Batch, (self.tasks,))
+
+
 class DaskDistributedBackend(ParallelBackendBase, AutoBatchingMixin):
-    MIN_IDEAL_BATCH_DURATION = 0.2
-    MAX_IDEAL_BATCH_DURATION = 1.0
+    MIN_IDEAL_BATCH_DURATION = 0.5
+    MAX_IDEAL_BATCH_DURATION = 5.0
 
-    def __init__(self, scheduler_host='127.0.0.1:8786', loop=None):
+    def __init__(self, scheduler_host='127.0.0.1:8786', scatter=None, loop=None):
         self.client = Client(scheduler_host, loop=loop)
+        if scatter is not None:
+            # Keep a reference to the scattered data to keep the ids the same
+            self._scatter = list(scatter)
+            scattered = self.client.scatter(scatter, broadcast=True)
+            self.data_to_future = {id(x): f for (x, f) in zip(scatter, scattered)}
+        else:
+            self._scatter = []
+            self.data_to_future = {}
         self.futures = set()
 
     def configure(self, n_jobs=1, parallel=None, **backend_args):
         return self.effective_n_jobs(n_jobs)
 
-    def effective_n_jobs(self, n_jobs=1):
+    def effective_n_jobs(self, n_jobs):
         return sum(self.client.ncores().values())
 
-    def apply_async(self, func, *args, **kwargs):
-        callback = kwargs.pop('callback', None)
-        kwargs['key'] = '%s-%s' % (joblib_funcname(func), uuid4().hex)
-        future = self.client.submit(func, *args, **kwargs)
+    def _to_func_args(self, func):
+        if not self.data_to_future:
+            return func, ()
+        args2 = []
+        lookup = dict(self.data_to_future)
+
+        def maybe_to_futures(args):
+            for x in args:
+                id_x = id(x)
+                if id_x in lookup:
+                    x = lookup[id_x]
+                    if type(x) is not itemgetter:
+                        x = itemgetter(len(args2))
+                        args2.append(lookup[id_x])
+                        lookup[id_x] = x
+                yield x
+
+        tasks = []
+        for f, args, kwargs in func.items:
+            args = list(maybe_to_futures(args))
+            kwargs = dict(zip(kwargs.keys(), maybe_to_futures(kwargs.values())))
+            tasks.append((f, args, kwargs))
+
+        if not args2:
+            return func, ()
+        return Batch(tasks), args2
+
+    def apply_async(self, func, callback=None):
+        key = '%s-%s' % (joblib_funcname(func), uuid4().hex)
```

Comment:
Should this add "%s-batch-%s" or something to signify that this is more than one function call?